lib/rubygems/requirement.rb

```diff
@@ -267,7 +267,17 @@ def to_s # :nodoc:
 
   def ==(other) # :nodoc:
     return unless Gem::Requirement === other
-    requirements == other.requirements
+
+    # An == check is always necessary
+    return false unless requirements == other.requirements
+
+    # An == check is sufficient unless any requirements use ~>
+    return true unless requirements.any? { |r| r.first == "~>" }
+
+    # If any requirements use ~> then we also need to compare the version
+    # strings for those requirements only (to check their precision is equal)
+    requirements.select { |r| r.first == "~>" }.map { |r| r.last.version_string } ==
+      other.requirements.select { |r| r.first == "~>" }.map { |r| r.last.version_string }
```

Comment:
`version_string` is already used by the `Version` class for `eql?`, so maybe if we use `eql?` here instead of `==`, we don't need to expose `#version_string`?

Reply:
We're comparing arrays here, and they might be in a different order.

`[1, 2].eql?([2, 1]) # => false`

Reply:
But I think [the PR]( https://github.com/rubygems/rubygems/pull/2230) that made this regression surface precisely ensures a consistent order when the class is initialized?

Reply:
The following change still passes tests, so if it's not correct, we should add more tests to prove it

```diff
diff --git a/lib/rubygems/requirement.rb b/lib/rubygems/requirement.rb
index 32dc824e..42399b79 100644
--- a/lib/rubygems/requirement.rb
+++ b/lib/rubygems/requirement.rb
@@ -276,8 +276,7 @@ class Gem::Requirement
 
     # If any requirements use ~> then we also need to compare the version
     # strings for those requirements only (to check their precision is equal)
-    requirements.select { |r| r.first == "~>" }.map { |r| r.last.version_string } ==
-      other.requirements.select { |r| r.first == "~>" }.map { |r| r.last.version_string }
+    requirements.select { |r| r.first == "~>" }.eql?(other.requirements.select { |r| r.first == "~>" })
   end
 
   private
```

Reply:
That's a good point. So could switch to:

```
requirements.select { |r| r.first == "~>" }.eql?(
  other.requirements.select { |r| r.first == "~>" }
)
```

Reply:
Haha, our comments crossed by that's identical code - I'll update. I did write a test for the case I was concerned about there.

Reply:
Simultaneous commenting! :smiley: